dpe/src/x509.rs

```diff
@@ -2227,6 +2235,286 @@ impl CertWriter<'_> {
     }
 }
 
+enum CertificateFormat {
+    X509,
+    Csr,
+}
+
+enum CertificateType {
+    Leaf,
+    Exported,
+}
+
+pub(crate) struct CreateDpeCertArgs<'a> {
+    pub handle: &'a ContextHandle,
+    pub locality: u32,
+    pub cdi_label: &'a [u8],
+    pub key_label: &'a [u8],
+    pub context: &'a [u8],
+}
+
+pub(crate) struct CreateDpeCertResult {
+    pub cert_size: u32,
+    pub pub_key: EcdsaPub,
+}
+
+fn get_dpe_measurement_digest(
+    dpe: &mut DpeInstance,
+    env: &mut DpeEnv<impl DpeTypes>,
+    handle: &ContextHandle,
+    locality: u32,
+) -> Result<Digest, DpeErrorCode> {
+    let parent_idx = dpe.get_active_context_pos(handle, locality)?;
+    let digest = dpe.compute_measurement_hash(env, parent_idx)?;
+    Ok(digest)
+}
+
+fn get_subject_name<'a>(
+    env: &mut DpeEnv<impl DpeTypes>,
+    pub_key: &'a EcdsaPub,
+    subj_serial: &'a mut [u8],
+) -> Result<Name<'a>, DpeErrorCode> {
+    env.crypto
+        .get_pubkey_serial(DPE_PROFILE.alg_len(), pub_key, subj_serial)?;
+
+    // The serial number of the subject can be at most 64 bytes
+    let truncated_subj_serial = &subj_serial[..64];
+
+    let subject_name = Name {
+        cn: DirectoryString::PrintableString(b"DPE Leaf"),
+        serial: DirectoryString::PrintableString(truncated_subj_serial),
+    };
+    Ok(subject_name)
+}
+
+fn get_tci_nodes(
```

Comment:
nit: maybe `get_tci_node_count`

Reply:
The `nodes` parameter is also an "out" parameter.

I'll refactor it to instead return a nodes buffer.

Reply:
Oh, got it. Well if this saves on stack space then feel free to leave it like this. I guess depends on how smart llvm is.

Reply:
I'll leave this for now since out params are not so nice.